### PR TITLE
Add Error Prone static analysis; fix some low hanging fruit

### PR DESCRIPTION
**Pull request by brandonweeks**
This PR adds [Error Prone](https://errorprone.info/) static analysis to the build using [gradle-errorprone-plugin](https://github.com/tbroyer/gradle-errorprone-plugin). I've also gone ahead and fixed a few classes of findings that didn't obviously impact cryptography. I'll open separate PRs or bugs for the cryptohtraphy impacting findings.

### CODE REVIEW

File: tls/src/main/java/org/bouncycastle/tls/SessionID.java [hunk 1]
```diff
@@ -4,7 +4,7 @@
 import org.bouncycastle.util.encoders.Hex;
 
 public final class SessionID
-    implements Comparable
+    implements Comparable<SessionID>
```
Comment (peterdettman) [inline tls/src/main/java/org/bouncycastle/tls/SessionID.java:7]: TLS currently still has to compile under Java 1.4 i.e. without generics. Please remove this file from the PR.

Comment (brandonweeks) [reply tls/src/main/java/org/bouncycastle/tls/SessionID.java]: Done.

File: core/src/main/java/org/bouncycastle/asn1/DateUtil.java [hunk 2]
```diff
@@ -64,7 +64,7 @@ static Date epochAdjust(Date date)
                 localeCache.put(locale, adj);
             }
 
-            if (adj != ZERO)
+            if (!adj.equals(ZERO))
```
Comment (peterdettman) [inline core/src/main/java/org/bouncycastle/asn1/DateUtil.java:67]: This is a false positive. Better remove this change and let us review the file separately.

**Comment by peterdettman** [comment]
Pending my last 2 review items, this LGTM.

**Review by peterdettman — CHANGES_REQUESTED** [review]

File: pkix/src/test/java/org/bouncycastle/dvcs/test/DVCSParseTest.java [hunk 3]
```diff
@@ -186,7 +188,7 @@ private static void validateArray(String type, Object[] result, Object[] expecte
 
         if (result.length != expected.length)
         {
-            fail("Different " + type + ": " + result + " while expected: " + expected);
+            fail("Different " + type + ": " + Arrays.toString(result) + " while expected: " + Arrays.toString(expected));
```
[the remark this answers is not in the document]
Comment (brandonweeks) [reply pkix/src/test/java/org/bouncycastle/dvcs/test/DVCSParseTest.java]: Done.

**Review by peterdettman — APPROVED** [review]
Thanks for making those changes. LGTM.

**Comment by dghgit** [comment]
Thanks for the patch. Now merged!

**Comment by dghgit** [comment]
Done as part of 1.72

**Pull request closed without merging**